### PR TITLE
Document error unref contract for iomgr cb's and transport/stream errors

### PR DESCRIPTION
**Pull request by vjpai**

### CODE REVIEW

File: src/core/lib/iomgr/closure.h [hunk 1]
```diff
@@ -42,7 +42,9 @@ typedef struct grpc_closure_list {
  *
  * \param arg Arbitrary input.
  * \param error GRPC_ERROR_NONE if no error occurred, otherwise some grpc_error
- *              describing what went wrong */
+ *              describing what went wrong.
+ *              Error contract: it is not the cb's job to unref this error;
```
Comment (markdroth) [inline src/core/lib/iomgr/closure.h]: FWIW, note that this is already documented here:

https://github.com/grpc/grpc/blob/master/doc/core/grpc-error.md

File: src/core/lib/transport/transport.h [hunk 2]
```diff
@@ -198,6 +198,8 @@ struct grpc_transport_stream_op_batch_payload {
         grpc_chttp2_grpc_status_to_http2_error. Send a RST_STREAM with this
         error. */
   struct {
+    // Error contract: the recipient of cancel_error must cause cancel_error
```
Comment (markdroth) [inline src/core/lib/transport/transport.h]: The term "recipient" is a bit unclear.  Perhaps this should say "the transport"?

Same thing in the two comments below.

File: src/core/lib/iomgr/closure.h (hunk 1 above)
Comment (vjpai) [inline src/core/lib/iomgr/closure.h]: Good point, but I feel that having it spelled out in the code is also useful.

Note that the other cases in this PR are not explicitly covered by the document: the document says that ownership transfers when an error is passed to a function (which sounds like an argument), but those other cases have a struct as their argument, one of whose members is the error.

File: src/core/lib/transport/transport.h (hunk 2 above)
Comment (vjpai) [inline src/core/lib/transport/transport.h]: That was my original thought, but this particular one can also go to grpc_transport_stream_op_finish_batch_with_failure which never makes it down to the `ext/transport` . But I guess a `lib/transport` is still part of the transport so it's still legitimate. I can change that.

Comment (vjpai) [inline src/core/lib/transport/transport.h]: Resolved this comment.

Comment (markdroth) [inline src/core/lib/transport/transport.h]: Sounds good.  You could also add a comment on `grpc_transport_stream_op_finish_batch_with_failure()` documenting the fact that it unrefs `cancel_error`.

Comment (vjpai) [inline src/core/lib/transport/transport.h]: Added a rhythmically-appropriate comment.

**Pull request merged**